tests/sharded_jit_test.py

```diff
@@ -0,0 +1,123 @@
+
+import jax
+import jax.numpy as np
+from jax import test_util as jtu
+from jax.interpreters import pxla
+from jax.interpreters.sharded_jit import sharded_jit
+from jax.interpreters.sharded_jit import PartitionSpec as P
+
+from jax.config import config
+config.parse_flags_with_absl()
+
+
+class ShardedJitTest(jtu.JaxTestCase):
+
+  def testBasic(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
+      raise SkipTest
+
+    @partial(sharded_jit, in_parts=(P(2, 1), P(2, 1)), out_parts=None)
+    def f(x, y):
+      return x + y
+
+    shape = (8, 8)
+    x = onp.arange(onp.prod(shape), dtype=np.float32).reshape(shape)
+    actual = f(x, x + 1)
+    expected = x + (x + 1)
+    self.assertAllClose(actual, expected, check_dtypes=False)
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), expected,
+                        check_dtypes=False)
+
+  def testPyTreeArgs(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
+      raise SkipTest
+
+    def f(a, b, c):
+      a1, a2 = a
+      c1, (c2, c3) = c
+      return a1 + a2 + b + c1 + c2 + c3
+
+    def _make_arg(*shape):
+      return onp.arange(onp.prod(shape)).reshape(shape)
+
+    a = (_make_arg(4, 4), _make_arg())
+    b = _make_arg(4, 4)
+    c = [_make_arg(), (_make_arg(4, 4), _make_arg(4, 4))]
+
+    in_parts = (None, P(2, 1), [None, P(2, 1)])
+    out_parts = P(2, 1)
+
+    result = sharded_jit(f, in_parts, out_parts)(a, b, c)
+    expected = f(a, b, c)
+
+    self.assertAllClose(result, expected, check_dtypes=False)
+    self.assertIsInstance(result, pxla.ShardedDeviceArray)
+    self.assertLen(result.device_buffers, 2)
+
+    in_parts = None
+    result = sharded_jit(f, in_parts, out_parts)(a, b, c)
+    self.assertAllClose(result, expected, check_dtypes=False)
+    self.assertIsInstance(result, pxla.ShardedDeviceArray)
+    self.assertLen(result.device_buffers, 2)
+
+  def testPyTreeOutputs(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
+      raise SkipTest
+
+    def f(x):
+      return x + 1, ((x + 2, x + 3), x + 4)
+
+    shape = (4, 4)
+    x = onp.arange(onp.prod(shape)).reshape(shape)
+    in_parts = (P(2, 1),)
+    out_parts = (P(2, 1), ((P(1, 2), None), P(2, 1)))
+
+    result = sharded_jit(f, in_parts, out_parts)(x)
+    expected = f(x)
+    self.assertAllClose(result, expected, check_dtypes=False)
+
+    out_parts = None
+    result = sharded_jit(f, in_parts, out_parts)(x)
+    self.assertAllClose(result, expected, check_dtypes=False)
+
+  def testAllReplicated(self):
+    if jtu.device_under_test() != "tpu":
+      raise SkipTest
+
+    @partial(sharded_jit, in_parts=None, out_parts=None)
+    def f(x):
+      return x + 1
+
+    result = f(1.)
+    self.assertEqual(result, 2.)
+    self.assertIsInstance(result, pxla.ShardedDeviceArray)
+    self.assertLen(result.device_buffers, 1)
```

Comment:
I do not understand in what way is this replicated? (the name of the test is `testAllReplicated`)

Reply:
The test name is referring to all inputs/outputs being specified as replicated, but you're right that this isn't meaningfully replicated. I renamed the test. This is kind of a funny case, but I think running a single "replica" is the only reasonable behavior (or possibly requiring that a separate `num_partitions` argument is provided, which I'm gonna add later to support this case with partitioned infeed).